### PR TITLE
Multi-aperture daophot ascii - clean up #3535

### PR DESCRIPTION
**Pull request by taldcroft**
This cleans up #3535 from @apodemus without changing any substantive code:

- PEP8 and style fixes
- Squash some commits (leaving key commits)
- Rebase on master
- Update changelog

cc: @hamogu

### CODE REVIEW

File: astropy/io/ascii/daophot.py [hunk 1]
```diff
@@ -12,15 +12,84 @@
...
+        stripped).
+        #N ID    XCENTER   YCENTER   MAG         MERR          MSKY           NITER
+        #U ##    pixels    pixels    magnitudes  magnitudes    counts         ##
+        #F %-9d  %-10.3f   %-10.3f   %-12.3f     %-14.3f       %-15.7g        %-6d
+        """
+        line_ids = ('#N', '#U', '#F')
+        coldef_dict = defaultdict(list)
+
+        # Function to strip identifier lines
+        stripper = lambda s: s[2:].strip(' \\')
```
Comment (hamogu) [inline astropy/io/ascii/daophot.py]: My linter complains when I assigned a lambda expression and says "if you assign it, use def."
Not sure if that's part of PEP8 though.

File: astropy/io/ascii/tests/test_read.py [hunk 2]
```diff
@@ -211,6 +211,14 @@ def test_daophot_multiple_aperture():
     assert np.all(table['RAPERT5'] == 23.3)  # assert all the 5th apertures are same 23.3
 
 
+def test_daophot_multiple_aperture2():
+    table = ascii.read('t/daophot4.dat', Reader=ascii.Daophot)
+    assert 'MAG15' in table.colnames  # MAG15 is one of the newly created column name
```
Comment (hamogu) [inline astropy/io/ascii/tests/test_read.py]: column nameS  :-;

File: astropy/io/ascii/misc.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+def groupmore(func=None, *its):
+    """Extends the itertools.groupby functionality to arbitrary number of iterators."""
+    if not func:
+        func = lambda x: x
+    its = sortmore(*its, key=func)
+    nfunc = lambda x: func(x[0])
+    zipper = itertools.groupby(zip(*its), nfunc)
+    unzipper = ((key, zip(*groups)) for key, groups in zipper)
+    return unzipper
```
Comment (hamogu) [inline astropy/io/ascii/misc.py]: This is too advanced for me.

File: astropy/io/ascii/misc.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
+"""A Collection of useful miscellaneous functions.
```
Comment (hamogu) [inline astropy/io/ascii/misc.py]: All these helpers are not ascii specific. Should they live in astropy.utils?

File: astropy/io/ascii/daophot.py [hunk 5]
```diff
@@ -12,15 +12,84 @@
...
 
 
 class DaophotHeader(core.BaseHeader):
-    comment = r'\s*#K'
-    aperture_values = ''
+    """Read the header from a file produced by the IRAF DAOphot routine."""
+
+    def __init__(self):
+        core.BaseHeader.__init__(self)
+        self.comment = r'\s*#K'
```
Comment (hamogu) [inline astropy/io/ascii/daophot.py]: I strongly prefer to define these things as class attributes, don't just assign them in `__init__`. That makes it harder to subclass this class (we had a earlier issue where I moved definitions like this from __init__ up one level (I just don't remember the number of that PR now).
The same applies to re_format, re_header_keyword etc.

**Comment by hamogu** [comment]
I've marked some stylistic comments, but there are too many zip, map, next in there for me to be comfortable. I'm sure it's fantastic code, but it's way over my head.
I'm usually thinking in numpy arrays not in iterators.

It seems to work, so I guess it's fine, but for a real code review you will have to find a more competent reviewer or wait until I have enough time to run this branch in a debugger line by line (and that won't happen before the feature freeze).

File: astropy/io/ascii/daophot.py (hunk 1 above)
Comment (embray) [inline astropy/io/ascii/daophot.py]: > Not sure if that's part of PEP8 though.

No, I don't really agree with that rule anyways.

File: astropy/io/ascii/daophot.py (hunk 5 above)
Comment (embray) [inline astropy/io/ascii/daophot.py]: Agree with @hamogu here.

Comment (taldcroft) [inline astropy/io/ascii/daophot.py]: Will fix.

File: astropy/io/ascii/misc.py (hunk 4 above)
Comment (taldcroft) [inline astropy/io/ascii/misc.py]: They would need their own unit tests and numpy-compliant docstrings, etc.  For now just leave them here.

**Pull request merged**